src/transport/raw/TCP.cpp

```diff
@@ -290,99 +279,93 @@ CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBuf
         {
-            // Open the receive window just enough to allow the remainder of the message to be received.
-            // This is necessary in the case where the message size exceeds the TCP window size to ensure
-            // the peer has enough window to send us the entire message.
-            uint16_t neededLen = static_cast<uint16_t>(messageSize - buffer->DataLength());
-            err                = endPoint->AckReceive(neededLen);
-            SuccessOrExit(err);
+            // This message is too long for upper layers, which require a contiguous buffer.
+            // TODO: Actual limit may be lower (spec issue #2119)
+            state->mReceived.Consume(kPacketSizeBytes);
+            state->mDiscardSize = messageSize;
+            ReturnErrorOnFailure(state->Discard());
+            return CHIP_ERROR_MESSAGE_TOO_LONG;
         }
-
-        // Buffer is incomplete and we cannot get more data
-        break;
+        // The subtraction will not underflow because we successfully read kPacketSizeBytes.
+        if (messageSize > (state->mReceived->TotalLength() - kPacketSizeBytes))
+        {
+            // We have not yet received the complete message.
+            return CHIP_NO_ERROR;
+        }
+        state->mReceived.Consume(kPacketSizeBytes);
+        ReturnErrorOnFailure(ProcessSingleMessage(peerAddress, state, messageSize));
     }
 
-exit:
-    if (!buffer.IsNull())
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TCPBase::ProcessSingleMessage(const PeerAddress & peerAddress, ActiveConnectionState * state, uint16_t messageSize)
+{
+    // We enter with `state->mReceived` containing at least one full message, perhaps in a chain.
+    // `state->mReceived->Start()` currently points to the message data.
+    // On exit, `state->mReceived` will have had `messageSize` bytes consumed, no matter what.
+    System::PacketBufferHandle message;
+    if (state->mReceived->DataLength() == messageSize)
+    {
+        // In this case, the head packet buffer contains exactly the message.
+        // This is common because typical messages fit in a network packet, and are delivered as such.
+        // Peel off the head to pass upstream, which effectively consumes it from `state->mReceived`.
+        message = state->mReceived.PopHead();
+    }
+    else
     {
-        // Incomplete processing will be retried
-        endPoint->PutBackReceivedData(std::move(buffer));
+        // The message is either longer or shorter than the head buffer.
+        // In either case, copy the message to a fresh linear buffer to pass upstream. We always copy, rather than provide
+        // a shared reference to the current buffer, in case upper layers manipulate the buffer in ways that would affect
+        // our use, e.g. chaining it elsewhere or reusing space beyond the current message.
+        message = System::PacketBufferHandle::New(messageSize, 0);
+        if (message.IsNull())
+        {
+            return CHIP_ERROR_NO_MEMORY;
```

Comment:
FWIW, I think this early return without calling `state->mReceived.Consume(messageSize);` is buggy: we have already consumed the `kPacketSizeBytes` that had the message length, so on OOM here we will come back later and reprocess things treating the first two payload bytes as a length.

This stops being an issue with the fix for #5438 so not a big deal.